ssh-server/main.go

```diff
@@ -66,11 +69,51 @@ func getConfiguration() (*ssh.ServerConfig, error) {
 		return nil, fmt.Errorf("creating signer: %v", err)
 	}
 
-	config := &ssh.ServerConfig{NoClientAuth: true}
+	config := &ssh.ServerConfig{
+		PublicKeyCallback: func(conn ssh.ConnMetadata, key ssh.PublicKey) (*ssh.Permissions, error) {
+			authorizedKeys, err := getAuthorizedKeys()
+			if err != nil {
+				return nil, fmt.Errorf("getting the authorized keys: %v", err)
+			}
+
+			if !authorizedKeys[string(key.Marshal())] {
+				return nil, errors.New("unauthorized")
+			}
+			return nil, nil
+		},
+	}
 	config.AddHostKey(signer)
 	return config, nil
 }
 
+func getAuthorizedKeys() (map[string]bool, error) {
+	req, err := http.NewRequest("GET", "http://metadata.google.internal/computeMetadata/v1/project/attributes/ssh-keys", nil)
+	if err != nil {
+		return nil, fmt.Errorf("building the request: %v", err)
+	}
+	req.Header.Set("Metadata-Flavor", "Google")
+	res, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return nil, fmt.Errorf("getting ssh keys from metadata server: %v", err)
+	}
+
+	keysBytes, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, fmt.Errorf("reading the ssh keys: %v", err)
+	}
+
+	authorizedKeys := map[string]bool{}
```

Comment:
This is OK, but more commonly use authorizedKeys := make(map[string]bool)

Reply:
done